File: EPI/8.12/typescript/8.12_test.ts
import {Node, isPalindrome} from './8.12';

declare function require(name: string): any;
const assert = require('assert');

function toList(values: number[]): Node {
    const nodes = values.map(value => ({value} as Node));
    nodes.forEach((node, i) => node.next = nodes[i + 1]);
    return nodes[0];
}

const tests: {list: number[], isPalindrome: boolean}[] = [
    {list: [1], isPalindrome: true},
    {list: [1, 1], isPalindrome: true},
    {list: [1, 2, 1], isPalindrome: true},
    {list: [1, 2], isPalindrome: false},
    {list: [1, 2, 2], isPalindrome: false},
    {list: [1, 5, 2, 3, 2, 5, 1], isPalindrome: true},
    {list: [1, 5, 2, 2, 5, 1], isPalindrome: true},
    {list: [1, 5, 2, 3, 5, 1], isPalindrome: false},
];

for (const test of tests) {
    const actual = isPalindrome(toList(test.list));
    assert.strictEqual(actual, test.isPalindrome, `For list ${test.list}, ` +
        `expected isPalindrome to be ${test.isPalindrome}, but was ${actual}`);
}

File: EPI/8.12/swift/8.12_test.swift
import Darwin

let tests = [
    (list: [0], isPalindrome: true),
    (list: [0, 1], isPalindrome: false),
    (list: [0, 0], isPalindrome: true),
    (list: [0, 1, 0, 3, 1, 1, 0], isPalindrome: false),
    (list: [0, 1, 0, 3, 0, 1, 0], isPalindrome: true),
]

func toList(_ values: [Int]) -> Node {
    let nodes = values.map(Node.init)
    for i in 0..<values.count - 1 {
        nodes[i].next = nodes[i + 1]
    }
    return nodes[0]
}

for test in tests {
    let actual = toList(test.list).isPalindrome()
    guard actual == test.isPalindrome else {
        print("For list \(test.list), expected isPalindrome to be \(test.isPalindrome), " +
            "but was \(actual)")
        exit(1)
    }
}
